### ISSUE TITLE
Add Sentry support for .NET debug symbols (portable pdbs) to unblock app distribution to customer devices without pdbs.

### ISSUE THREAD

**Issue by asbjornu** [getsentry/sentry-dotnet#193]
As in getsentry/raven-objc#11, we need to add support for debug symbols in the C# client. This requires back-end support from @dcramer and his team, but as soon as that is in place, we should be able to upload `.pdb` files and have them attached as artifacts to a release.

**Comment by PureKrome** [comment]
even better @bruno-garcia is there's really nice SourceLink ability to step into code really easy, in dotnet core.
- https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink
- https://github.com/dotnet/sourcelink

simple.
easy to setup.
works like a treat!

**Comment by bruno-garcia** [comment]
Hi @PureKrome! Thanks for chipping in!

On the SDK side, we've got SourceLink support from beginning:

https://github.com/getsentry/sentry-dotnet/commit/fe4f617d6c9600105e45f6542c4c90538f2269b6

That is to help you debug through the SDK code if you need, and also for a tool like Sentry to display source context if it's involved in a crash. Which is what I think you meant.

The goal of this issue is to allow customers upload their .NET `.pdb` to Sentry (which we'd need anyway in order to access the sourcelink data, unless processed/sent by the client which means the client has the PDBs anyway).

This would allow what many desktop apps would want: not release their PDBs with their binaries. Sometimes because of the size constraints (less of an issue with portable PDBs which are ~30% of the size of the binary) but also to make reverse engineering harder. 

Right now, if you don't have the PDBs in the client, the stack traces would actually have no information and we'd need to symbolicate in the processing pipeline in Sentry. We can't do that right now for .NET PDBs ([we've recently added support to Native PDBs though](https://blog.sentry.io/2019/05/23/native-crash-reporting-symbol-servers-pdbs-sdk-c-c-plus-plus)).

Sourcelink would allow us to show the code around the affected lines but we'd still need the PDB to symbolicate and also to get the sourcelink data which is embedded in the PDB. And now there are two types of PDBs (old Windows PDB and Portable PDB) so we need to deal with that too :)

**Comment by Green7** [comment]
Any status on this?
I see that I can upload pdb files but my stack trace still have no additional information ...

**Comment by bruno-garcia** [comment]
@Green7 Sorry but this issue was not addressed yet.
I'm interested in more details of your use-case btw. 

A few things we'd like to understand are:

1. What type and version of .NET: .NET Framework 4.5, .NET Core 2.0
2. What type of PDB do you use (full, portable).
3. Do you use any obfuscator, if yes, which one?
4. What's the reason not to deploy PDBs with the app. Size constraints? Other?
5. If it was possible to upload your PDBs to Sentry, would be willing to [embed the source code in the PDB](https://github.com/dotnet/roslyn/issues/12625) to get source context (your actual source code around the stack trace lines) in the Sentry dashboard.

**Comment by Green7** [comment]
1. Net Framework 4.6
2. Full PDB
3. Not now, but I want to use Crypto Obfuscator For .Net  #186 
4. Size and security reasons
5. This is a nice feature but it is not essential for me. The most important is stack trace along with the corresponding line numbers of the source code. With this information I can check the source context in VCS.

**Comment by Green7** [comment]
Any news ? 
The stack is useless without pdb files but we don't want to distribute pdb files to our users.
Is there any time line for handling this?

**Comment by bruno-garcia** [comment]
@Green7 Unfortunately we don't have this work planned at the moment. We'll be looking into what's next for .NET in the upcoming months though.

**Comment by bruno-garcia** [comment]
@Green7 We need some data to prioritize this. You mentioned .NET Framework 4.6 and full PDB.
Is this a WPF or WinForms or else?
Also, do you plan to move to 4.6.1 anytime soon or .NET Core 3 now that it supports WinForms/WPF?

**Comment by Green7** [comment]
WinForms. For now we must stay on .Net Framework 4.6 because we still have customers working on Windows Server 2008 or older.

**Comment by bruno-garcia** [comment]
Now with WinForms and WPF going to .NET Core, I hope we'll set sometime aside to invest in it here at Sentry.

Note that this SDK doesn't support .NET Framework 4.6. Only 4.6.1+. Sentry still has its [Raven SDK](https://github.com/getsentry/raven-csharp) which supports .NET all the way to 3.5. We maintain it for bugfixes but no new feature are added there though.

Also, I'm curious what type of WinForms app would run on a server. Could you share some info?

**Comment by Green7** [comment]
We using Raven SDK. We create a computer management application and some customers use it on servers.

**Comment by bddckr** [comment]
Here's my use case:
* Various .NET Core (3.0+) CLI apps.
* A .NET Core (3.0+) WinForms app.
* A Xamarin.Android app.

* No obfuscator in use, but this might change.
* Size is the main concern for us.
* No interest in embedding the source code into the PDBs currently, but this might change.

**Comment by Lonli-Lokli** [comment]
Why PDBs are required to be uploaded to Sentry? As for me if we have AWS ECS endpoint (or any other docker-based) PDB can be included in docker image.

**Comment by bruno-garcia** [comment]
@Lonli-Lokli I'm unaware of a legitimate use case of not publishing server apps with PDBs.
This issue is relevant to client apps (like WPF, WinForms) where the app developer is concerned about either size of the final bundle (less of an issue today with Portable PDBs which are tiny) or more commonly, reverse engineering their code and cloning the app.

**Comment by PureKrome** [comment]
Out of interest - is this issue not as relevant any more, now that we have Sourcelink integration?

**Comment by bruno-garcia** [comment]
@PureKrome sourcelink will help point you to the source (commit hash and path) but it won't help .NET give you the line numbers of a function if there's no `pdb` around. I believe sourcelink data is stored in the `pdb` although it's possible it can be stored in the `dll` instead

**Comment by MeikTranel** [comment]
Any update on this? Does sentry support managed pdb / portable pdbs by now?

**Comment by bruno-garcia** [comment]
This isn't something we've added yet but I'm bringing this conversation internally to see if we can start planning this.
@MeikTranel could you please share your use case? Is this for Unity? Desktop apps? Do you use obfuscating tools? Do you use an internal NuGet server or how do you track your pdbs?

**Comment by MeikTranel** [comment]
I have use cases for both desktop and server apps with sourcelink and non sourcelink pens, full and portable symbols as well as obfuscation where we would have to do a relay filter to deobfuscate as well as non obfuscated stack frames

**Comment by bruno-garcia** [comment]
@MeikTranel It's unclear that we'll cover all of that, of what combinations. To clarify: Obfuscation is a separate topic (tracked here: https://github.com/getsentry/sentry-dotnet/issues/186 please add details there like which obfuscation tool you use).

This issue will cover the ability to upload portable PDBs (not Windows PDBs) to Sentry, and .NET stack traces would get symbolicated by it. This will serve Mobile/Desktop use cases (given for server loads it's simpler to embed the pdbs or have them side-by-side since they are tiny).

**Comment by PureKrome** [comment]
@bruno-garcia can you update the thread topic to be more detailed to highlight that this is for mobile/desktop and not about generating pdb's for the sentry nugets (blush) 😄

**Comment by bddckr** [comment]
Just chiming in here with my team's use case:
* Windows, macOS (Xamarin.Mac), Android (Xamarin.Android) apps (all soon to be on .NET 6)
    * (We've been using stock Sentry .NET for this, not using any Xamarin-specifics or anything here since we share most of our code among the platforms.)
* `DebugType` is set to `portable` ✔️

Supporting `portable` seems reasonable, considering that it's the only supported format starting in .NET 6:
> `portable` is the recommended setting for all projects from now on. The older `full` and `pdbonly` settings are for older Windows-specific file formats. .NET 6 and higher will not support those older formats.
> https://docs.microsoft.com/xamarin/android/release-notes/11/11.2#deprecation-of-debugtype-full-and-pdbonly

**Comment by bruno-garcia** [comment]
> @bruno-garcia can you update the thread topic to be more detailed to highlight that this is for mobile/desktop and not about generating pdb's for the sentry nugets (blush) 😄

I wasn't aware there was confusion.
To clarify: The Sentry SDK for .NET built in this repo has sourcelink and pdbs available since its inception.
This issue discusses a feature at Sentry where you can upload to Sentry your applications **pdb**s (usually a concern of client apps where you don't want to make pdbs available to the customers). And Sentry will use those pdbs to show your proper stack traces with line numbers and file paths.

> Supporting portable seems reasonable, considering that it's the only supported format starting in .NET 6:

Good point. portable pdbs have been available on Mono, .NET Framework (and obviously .NET Core) for many years now and I'm unaware of blockers for anyone to adopt even when using older versions of the framework.

**Comment by triforcely** [comment]
@bruno-garcia would that allow for uploading of "source bundles" as well? I'm currently shipping serverside code with PDBs, so I get line numbers in stack traces but to have the full experience my team would like to see source code as well. AFAIK I can't use sentry-cli to upload source bundles because it doesn't work with .NET PDBs.

**Comment by bruno-garcia** [comment]
That's correct, source bundles isn't support yet. We are planning on doing stacktrace linking though: https://docs.sentry.io/product/integrations/source-code-mgmt/github/#stack-trace-linking

**Comment by triforcely** [comment]
> That's correct, source bundles isn't support yet. We are planning on doing stacktrace linking though: https://docs.sentry.io/product/integrations/source-code-mgmt/github/#stack-trace-linking

That would be great. Will bitbucket be supported?

**Comment by Lonli-Lokli** [comment]
> > That's correct, source bundles isn't support yet. We are planning on doing stacktrace linking though: https://docs.sentry.io/product/integrations/source-code-mgmt/github/#stack-trace-linking
> 
> That would be great. Will bitbucket be supported?

Its already supported https://docs.sentry.io/product/integrations/source-code-mgmt/

**Comment by bruno-garcia** [comment]
If you're subscribed to this, not just PDB support but source context support landed:

![image](https://user-images.githubusercontent.com/1633368/216153521-9a56a4ad-0192-4291-aa44-208065f4b021.png)

https://twitter.com/brungarc/status/1620877512841777152
